### PR TITLE
Add new endpoints and tests cases to upgrade and get upgrade results of multiple agents

### LINKED ISSUE #5536: [WPK][API] Add new endpoints to upgrade and get upgrade results of multiple agents
Hello team,

This issue is based on https://github.com/wazuh/wazuh/issues/5387. 

## Description

With the refactor and changes to the upgrade process, the new Wazuh upgrade module allows to receive a petition to upgrade multiple agents at once. We should add new API endpoints to allow receiving a list of agents to ugprade, or all of them. We should  add the same endpoint with custom upgrade and upgrade results.

## Design

The endpoints should have the following structure:

Path: 

- `PUT /agents/upgrade`

Query parameters: 
- `pretty`
- `wait_for_complete`
- `list_agents`
- `wpk_repo`
- `upgrade_version`
- `use_http`
- `force`

Path: 

- `PUT /agents/upgrade_custom`

Query parameters: 
- `pretty`
- `wait_for_complete`
- `list_agents`
- `file_path`
- `installer`

Path: 

- `GET /agents/upgrade_result`

Query parameters: 
- `pretty`
- `wait_for_complete`
- `list_agents`

## Testing

We should add new endpoint test integration cases for the 3 new endpoints to:
-  [test_agents_PUT_endpoints.tavern.yaml](https://github.com/wazuh/wazuh/blob/develop/api/test/integration/test_agents_PUT_endpoints.tavern.yaml)
-  [test_rbac_black_agents_endpoints.tavern.yaml](https://github.com/wazuh/wazuh/blob/develop/api/test/integration/test_rbac_black_agents_endpoints.tavern.yaml)
- [test_rbac_white_agents_endpoints.tavern.yaml](https://github.com/wazuh/wazuh/blob/develop/api/test/integration/test_rbac_white_agents_endpoints.tavern.yaml)

## Tasks

- [x] Update spec.yaml
- [x] Update/Add agents controllers
- [x] Add and run new test integration cases
- [x] Update changes.md

### PR DESCRIPTION
**Pull request by mcarmona99**
**Related issue**
#5536 

Hello team,

This PR closes #5536 . In this PR I have added new API endpoints to allow receiving a list of agents to upgrade, or all of them. Endpoints upgrade custom and get upgrade results for a list of agents (or all of them) have been added, too.
Endpoints:
- PUT /agents/upgrade
- PUT /agents/upgrade_custom
- GET /agents/upgrade_result

New test cases added to:
- test_agents_PUT_endpoints.tavern.yaml
- test_rbac_black_agents_endpoints.tavern.yaml
- test_rbac_white_agents_endpoints.tavern.yaml

Changes on spec.yaml, agents_controller.py, changes.md and on the three integration tests mentioned above.

### CODE REVIEW

File: api/api/controllers/agents_controller.py [hunk 1]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
     return web.json_response(data=data, status=200, dumps=prettify if pretty else dumps)
 
 
+async def put_upgrade_multiple_agents(request, list_agents, pretty=False, wait_for_complete=False, wpk_repo=None,
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:468]: Remove old functions (407\~465, 557\~579) and remove the word "multiple"

File: api/api/controllers/agents_controller.py [hunk 2]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
...
+                          wait_for_complete=True,  # Force wait_for_complete until timeout problems are resolved
+                          logger=logger,
+                          rbac_permissions=request['token_info']['rbac_policies']
+                          )
+    data = raise_if_exc(await dapi.distribute_function())
+
+    return web.json_response(data=data, status=200, dumps=prettify if pretty else dumps)
+
+
+async def put_upgrade_multiple_custom_agents(request, list_agents, pretty=False, wait_for_complete=False,
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:501]: Remove "multiple"

File: api/api/controllers/agents_controller.py [hunk 3]
```diff
@@ -515,6 +579,31 @@ async def get_agent_upgrade(request, agent_id, timeout=3, pretty=False, wait_for
     return web.json_response(data=data, status=200, dumps=prettify if pretty else dumps)
 
 
+async def get_agent_upgrade_multiple(request, list_agents, timeout=3, pretty=False, wait_for_complete=False):
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:582]: Remove "multiple"

File: api/api/controllers/agents_controller.py [hunk 4]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
     return web.json_response(data=data, status=200, dumps=prettify if pretty else dumps)
 
 
+async def put_upgrade_multiple_agents(request, list_agents, pretty=False, wait_for_complete=False, wpk_repo=None,
+                                      version=None,
+                                      use_http=False, force=False):
+    """Upgrade agents using a WPK file from online repository.
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:471]: Change the docstring to numpydoc

File: api/api/controllers/agents_controller.py [hunk 5]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
...
+                          )
+    data = raise_if_exc(await dapi.distribute_function())
+
+    return web.json_response(data=data, status=200, dumps=prettify if pretty else dumps)
+
+
+async def put_upgrade_multiple_custom_agents(request, list_agents, pretty=False, wait_for_complete=False,
+                                             file_path=None,
+                                             installer=None):
+    """Upgrade agents using a local WPK file.'.
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:504]: Change the docstring to numpydoc

File: api/api/controllers/agents_controller.py [hunk 6]
```diff
@@ -515,6 +579,31 @@ async def get_agent_upgrade(request, agent_id, timeout=3, pretty=False, wait_for
     return web.json_response(data=data, status=200, dumps=prettify if pretty else dumps)
 
 
+async def get_agent_upgrade_multiple(request, list_agents, timeout=3, pretty=False, wait_for_complete=False):
+    """Get upgrade results from agents.
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:583]: Change the docstring to numpydoc

File: api/api/controllers/agents_controller.py [hunk 7]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
...
+    """
+    f_kwargs = {'list_agents': list_agents,
+                'file_path': file_path,
+                'installer': installer}
+
+    dapi = DistributedAPI(f=agent.upgrade_multiple_custom_agents,
+                          f_kwargs=remove_nones_to_dict(f_kwargs),
+                          request_type='distributed_master',
+                          is_async=False,
+                          wait_for_complete=True,  # Force wait_for_complete until timeout problems are resolved
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:522]: Now we can use the wait_for_complete parameter

File: api/api/controllers/agents_controller.py [hunk 8]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
...
+                'wpk_repo': wpk_repo,
+                'version': version,
+                'use_http': use_http,
+                'force': force}
+
+    dapi = DistributedAPI(f=agent.upgrade_multiple_agents,
+                          f_kwargs=remove_nones_to_dict(f_kwargs),
+                          request_type='distributed_master',
+                          is_async=False,
+                          wait_for_complete=True,  # Force wait_for_complete until timeout problems are resolved
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:492]: Now we can use the wait_for_complete_parameter

File: api/api/controllers/agents_controller.py [hunk 9]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
...
+    :param force: Force upgrade.
+    :return: ApiResponse
+    """
+    f_kwargs = {'list_agents': list_agents,
+                'wpk_repo': wpk_repo,
+                'version': version,
+                'use_http': use_http,
+                'force': force}
+
+    dapi = DistributedAPI(f=agent.upgrade_multiple_agents,
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:488]: Change upgrade_multiple_agents for upgrade_agents (new function in framework side)

File: api/api/controllers/agents_controller.py [hunk 10]
```diff
@@ -464,6 +465,69 @@ async def put_upgrade_custom_agent(request, agent_id, pretty=False, wait_for_com
...
+    :param file_path: Path to the WPK file. The file must be on a folder on the Wazuh's installation directory
+    (by default, <code>/var/ossec</code>).
+    :type installer: str
+    :return: ApiResponse
+    """
+    f_kwargs = {'list_agents': list_agents,
+                'file_path': file_path,
+                'installer': installer}
+
+    dapi = DistributedAPI(f=agent.upgrade_multiple_custom_agents,
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:518]: Change upgrade_multiple_agents for upgrade_agents (new function in framework side)

File: api/api/controllers/agents_controller.py [hunk 11]
```diff
@@ -515,6 +579,31 @@ async def get_agent_upgrade(request, agent_id, timeout=3, pretty=False, wait_for
...
+    :param pretty: Show results in human-readable format
+    :param wait_for_complete: Disable timeout response
+    :param list_agents: Agent ID. All posible values since 000 onwards.
+    :param timeout: Seconds to wait for the agent to respond.
+    :return: ApiResponse
+    """
+    f_kwargs = {'list_agents': list_agents,
+                'timeout': timeout}
+
+    dapi = DistributedAPI(f=agent.get_upgrade_multiple_result,
```
Comment (AdriiiPRodri) [inline api/api/controllers/agents_controller.py:594]: Remove "multiple"

File: api/api/spec/changes.md [hunk 12]
```diff
@@ -134,6 +134,15 @@ This agent will use **any** as IP.
 ### PUT     /agents/{agent_id}/upgrade
```
Comment (AdriiiPRodri) [inline api/api/spec/changes.md:134]: Delete this endpoint

File: api/api/spec/spec.yaml [hunk 13]
```diff
@@ -5769,6 +5769,111 @@ paths:
         '429':
           $ref: '#/components/responses/TooManyRequestsResponse'
 
+  /agents/upgrade:
```
Comment (AdriiiPRodri) [inline api/api/spec/spec.yaml:5772]: Remove old endpoints (5661/~5770)

File: api/api/spec/spec.yaml [hunk 14]
```diff
@@ -5769,6 +5769,111 @@ paths:
...
+    put:
+      tags:
+        - agents
+      summary: "Upgrade multiple agents using online repository"
+      description: "Upgrade multiple agents using a WPK file from online repository"
+      operationId: api.controllers.agents_controller.put_upgrade_multiple_agents
```
Comment (AdriiiPRodri) [inline api/api/spec/spec.yaml:5778]: Remove "multiple"

File: api/api/spec/spec.yaml [hunk 15]
```diff
@@ -5769,6 +5769,111 @@ paths:
...
+        '429':
+          $ref: '#/components/responses/TooManyRequestsResponse'
+
+  /agents/upgrade_custom:
+    put:
+      tags:
+        - agents
+      summary: "Upgrade agents using a custom file"
+      description: "Upgrade the agents using a local WPK file"
+      operationId: api.controllers.agents_controller.put_upgrade_multiple_custom_agents
```
Comment (AdriiiPRodri) [inline api/api/spec/spec.yaml:5813]: Remove "multiple"

File: api/api/spec/spec.yaml [hunk 16]
```diff
@@ -5769,6 +5769,111 @@ paths:
...
+        '429':
+          $ref: '#/components/responses/TooManyRequestsResponse'
+
+  /agents/upgrade_result:
+    get:
+      tags:
+        - agents
+      summary: "Get upgrade results from agents"
+      description: "Return the upgrade results after updating agents"
+      operationId: api.controllers.agents_controller.get_agent_upgrade_multiple
```
Comment (AdriiiPRodri) [inline api/api/spec/spec.yaml:5846]: Remove "multiple"

**Review by AdriiiPRodri — CHANGES_REQUESTED** [review]

**Comment by mcarmona99** [comment]
The changes made in this PR have been merged into the branch `dev-5537-refactor_framework_upgrade_module` and will be merged into master in the PR: #5979 .

**Pull request closed without merging**